### PR TITLE
[Fix] Add delay property for checkpointing, refactor loading checkpoint (DeepSpeed Checkpointing Fix 1/n) 

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Related to #8397. 

This PR introduces a property to the training type plugin that allows the plugin to delay loading the checkpoint till after pre-dispatch, and moves loading the checkpoint file (`resume_start`) into the same block of logic.

After discussions with @awaelchli I couldn't unfortunately come up with another method to load via the deepspeed engine but to introduce this branch into the `run` function. This branch will allow us to call `deepspeed_engine.load_checkpoint` when loading the checkpoint file.

**This PR relys on #8347 which will not be included till 1.5, so should skip this from 1.4.x**. I'll also to a combo CHANGELOG update in #8397 instead of here, as on it's own its confusing!

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -510,6 +510,15 @@ def setup_optimizers_in_pre_dispatch(self) -> bool:
         """
         return self.training_type_plugin.setup_optimizers_in_pre_dispatch
 
+    @property
+    def restore_checkpoint_after_pre_dispatch(self) -> bool:
+        """
+        Override to delay restoring from checkpoint till after predispatch.
+        This is useful when the plugin requires all the setup hooks to run before loading checkpoint.
+        Returns: If true, restore checkpoint after pre_dispatch.
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/accelerator.py:518]:
```suggestion
        Override to delay restoring from checkpoint till after pre-dispatch.
        This is useful when the plugin requires all the setup hooks to run before loading checkpoint.
        Returns: If true, restore checkpoint after pre_dispatch.
```
not sure about spelling, maybe it's correct :) )

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -838,6 +836,23 @@ def tune(
...
+        self.checkpoint_connector.restore_datamodule()
+        self.checkpoint_connector.restore_model()
+        # restore callback states
+        self.checkpoint_connector.restore_callbacks()
+
+    def _load_checkpoint_weights(self):
+        # only one process running at this point for TPUs, as spawn isn't triggered yet
+        # todo: move this logic internally within the barrier.
+        if not self._device_type == DeviceType.TPU:
+            self.training_type_plugin.barrier()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:852]:
```suggestion
            self.accelerator.barrier()
```

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -838,6 +836,23 @@ def tune(
...
+        rank_zero_info(f"Loading checkpoint from {self._ckpt_path}")
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:853]: the message should probably be more specific here that we only restore the weights. 
it's not restoring training, callbacks, etc. right?
```suggestion
        rank_zero_info(f"Loading model weights from checkpoint at {self._ckpt_path}")
```

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -838,6 +836,23 @@ def tune(
 
         return result
 
+    def _restore_checkpoint(self) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:839]: I believe we need to find a better name for this, as it does not restore the full checkpoint. 
Context: We separated the restoring of model and training state on purpose because it needs to happen at different stages of the whole fit and accelerator setup. 

```suggestion
    def _restore_modules_and_callbacks(self) -> None:
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 5]
```diff
@@ -235,6 +235,15 @@ def setup_optimizers_in_pre_dispatch(self) -> bool:
         """
         return False
 
+    @property
+    def restore_checkpoint_after_pre_dispatch(self) -> bool:
+        """
+        Override to delay restoring from checkpoint till after predispatch.
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:241]:
```suggestion
        Override to delay restoring from checkpoint till after pre-dispatch.
```

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
Comment (awaelchli) [inline pytorch_lightning/accelerators/accelerator.py:518]:
```suggestion

        Returns: 
            If true, restore checkpoint after pre_dispatch.
```
for correct rendering, this whitespace is needed afaik

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 6]
```diff
@@ -235,6 +235,15 @@ def setup_optimizers_in_pre_dispatch(self) -> bool:
...
+        This is useful when the plugin requires all the setup hooks to run before loading checkpoint.
+        Returns: If true, restore checkpoint after pre_dispatch.
+        """
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:244]:
```suggestion

        Returns: 
            If true, restore checkpoint after pre_dispatch.
        """
```

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (SeanNaren) [reply pytorch_lightning/trainer/trainer.py:853]: Yep agreed, just pulled all this portion from the previous written code, this is more informative!

**Review by tchaton — APPROVED** [review]
LGTM ! I don't see any obvious other solution either.

**Review by awaelchli — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]
